code/io/joy-sdl.cpp

```diff
@@ -554,11 +576,44 @@ namespace joystick
 		Assertion(hat < numHats(), "SDL event contained invalid hat index!");
 
 		auto hatpos = convertSDLHat(evt.value);
-		_hat[hat].DownTimestamp = (hatpos != HAT_CENTERED) ? timer_get_milliseconds() : -1;
-		_hat[hat].Value = hatpos;
 
-		if (hatpos != HAT_CENTERED)
-		{
+		// Reset timestamps
+		for (auto i = 0; i < JOY_NUM_HAT_POS; ++i) {
+			_hat[hat].DownTimestamp[i] = -1;
```

Comment:
I think it would be more desirable to keep the `DownTimestamp` values in some situations; e.g. when moving from e.g. `HAT_LEFT` to `HAT_LEFTUP`, keeping the `HAT_LEFT` timestamp because, well, the hat hasn't stopped pointing left.